File: agent/capabilities/persistence_detection.py
"""
Persistence Detection Capability
Scans for common persistence mechanisms on Windows and Linux.
"""
import os
import sys
import logging
import subprocess
from typing import Dict, Any, List
from .base import BaseCapability

logger = logging.getLogger(__name__)

# Commands that suggest malicious persistence when found in a scheduled/startup entry
_SUSPICIOUS_KEYWORDS = [
    "powershell", "cmd.exe", "wscript", "cscript", "mshta", "rundll32",
    "regsvr32", "certutil", "bitsadmin", "curl", "wget", "nc ", "ncat",
    "python", "perl", "ruby", "base64", "encodedcommand", "-enc ",
    "/tmp/", "\\temp\\", "\\appdata\\local\\temp\\",
]

def _risk_score(command: str) -> str:
    cmd_lower = (command or "").lower()
    hits = sum(1 for kw in _SUSPICIOUS_KEYWORDS if kw in cmd_lower)
    if hits >= 2:
        return "Critical"
    if hits == 1:
        return "High"
    if "appdata" in cmd_lower or "temp" in cmd_lower:
        return "Medium"
    return "Low"


class PersistenceDetectionCapability(BaseCapability):

    @property
    def capability_id(self) -> str:
        return "persistence_detection"

    @property
    def capability_name(self) -> str:
        return "Persistence Detection"

    def collect(self) -> Dict[str, Any]:
        findings = []
        try:
            if os.name == "nt":
                findings = self._scan_windows()
            else:
                findings = self._scan_linux()
        except Exception as e:
            logger.error("Persistence scan failed: %s", e)
            return {"error": str(e), "findings": []}

        critical = sum(1 for f in findings if f["severity"] == "Critical")
        high     = sum(1 for f in findings if f["severity"] == "High")
        return {
            "findings": findings,
            "count": len(findings),
            "critical_count": critical,
            "high_count": high,
            "platform": os.name,
        }

    # ── Windows ───────────────────────────────────────────────────────────────

    def _scan_windows(self) -> List[Dict[str, Any]]:
        findings = []
        try:
            import winreg
        except ImportError:
            return findings

        run_keys = [
            (winreg.HKEY_CURRENT_USER,  r"Software\Microsoft\Windows\CurrentVersion\Run"),
            (winreg.HKEY_CURRENT_USER,  r"Software\Microsoft\Windows\CurrentVersion\RunOnce"),
            (winreg.HKEY_CURRENT_USER,  r"Software\Microsoft\Windows\CurrentVersion\RunServices"),
            (winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows\CurrentVersion\Run"),
            (winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows\CurrentVersion\RunOnce"),
            (winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows\CurrentVersion\RunServices"),
            (winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows NT\CurrentVersion\Winlogon"),
            (winreg.HKEY_LOCAL_MACHINE, r"System\CurrentControlSet\Services"),
            # WOW64 variants
            (winreg.HKEY_CURRENT_USER,  r"Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Run"),
            (winreg.HKEY_LOCAL_MACHINE, r"Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Run"),
        ]

        for hkey, path in run_keys:
            hive = "HKCU" if hkey == winreg.HKEY_CURRENT_USER else "HKLM"
            try:
                with winreg.OpenKey(hkey, path) as key:
                    i = 0
                    while True:
                        try:
                            name, value, _ = winreg.EnumValue(key, i)
                            severity = _risk_score(value)
                            findings.append({
                                "type": "Registry Run Key",
                                "location": f"{hive}\\{path}",
                                "name": name,
                                "command": value,
                                "severity": severity,
                            })
                            i += 1
                        except OSError:
                            break
            except Exception:
                continue

        # Startup folders
        startup_paths = [
            os.path.join(os.environ.get("APPDATA", ""), r"Microsoft\Windows\Start Menu\Programs\Startup"),
            r"C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Startup",
        ]
        for path in startup_paths:
            if os.path.isdir(path):
                for item in os.listdir(path):
                    full = os.path.join(path, item)
                    findings.append({
                        "type": "Startup Folder",
                        "location": path,
                        "name": item,
                        "command": full,
                        "severity": _risk_score(item),
                    })

        # Scheduled tasks via schtasks
        findings.extend(self._scan_schtasks())

        return findings

    def _scan_schtasks(self) -> List[Dict[str, Any]]:
        findings = []
        try:
            output = subprocess.check_output(
                ["schtasks", "/query", "/fo", "CSV", "/v"],
                stderr=subprocess.DEVNULL,
                timeout=30,
            ).decode(errors="replace")
            for line in output.splitlines():
                if "Run As User" in line or "Task To Run" not in line:
                    continue
                parts = line.split(",")
                if len(parts) >= 9:
                    task_name = parts[0].strip().strip('"')
                    task_cmd  = parts[8].strip().strip('"')
                    if task_name and task_cmd and task_cmd.upper() != "TASK TO RUN":
                        severity = _risk_score(task_cmd)
                        if severity in ("Critical", "High", "Medium"):
                            findings.append({
                                "type": "Scheduled Task",
                                "location": "Windows Task Scheduler",
                                "name": task_name,
                                "command": task_cmd,
                                "severity": severity,
                            })
        except Exception as e:
            logger.debug("schtasks scan failed: %s", e)
        return findings

    # ── Linux ─────────────────────────────────────────────────────────────────

    def _scan_linux(self) -> List[Dict[str, Any]]:
        findings = []

        # systemd services
        for systemd_dir in ["/etc/systemd/system", "/lib/systemd/system", "/usr/lib/systemd/system"]:
            if os.path.isdir(systemd_dir):
                for item in os.listdir(systemd_dir):
                    if item.endswith(".service"):
                        full = os.path.join(systemd_dir, item)
                        cmd_line = self._read_service_exec(full)
                        findings.append({
                            "type": "systemd Service",
                            "location": full,
                            "name": item,
                            "command": cmd_line,
                            "severity": _risk_score(cmd_line),
                        })

        # user crontab
        try:
            output = subprocess.check_output(["crontab", "-l"], stderr=subprocess.STDOUT, timeout=5).decode()
            for line in output.splitlines():
                if line.strip() and not line.startswith("#"):
                    findings.append({
                        "type": "Crontab Entry",
                        "location": "User Crontab",
                        "name": "Cron Job",
                        "command": line.strip(),
                        "severity": _risk_score(line),
                    })
        except Exception:
            pass

        # /etc/cron.d/ and /etc/cron.* directories
        for cron_dir in ["/etc/cron.d", "/etc/cron.hourly", "/etc/cron.daily",
                         "/etc/cron.weekly", "/etc/cron.monthly"]:
            if os.path.isdir(cron_dir):
                for item in os.listdir(cron_dir):
                    full = os.path.join(cron_dir, item)
                    if os.path.isfile(full):
                        try:
                            content = open(full).read()
                            for line in content.splitlines():
                                if line.strip() and not line.startswith("#"):
                                    findings.append({
                                        "type": "Cron File",
                                        "location": full,
                                        "name": item,
                                        "command": line.strip(),
                                        "severity": _risk_score(line),
                                    })
                        except Exception:
                            pass

        # /etc/profile.d/ scripts
        if os.path.isdir("/etc/profile.d"):
            for item in os.listdir("/etc/profile.d"):
                full = os.path.join("/etc/profile.d", item)
                if os.path.isfile(full):
                    findings.append({
                        "type": "Profile Script",
                        "location": full,
                        "name": item,
                        "command": f"(loaded at login)",
                        "severity": "Low",
                    })

        # User shell config files
        for shell_file in [
            os.path.expanduser("~/.bashrc"),
            os.path.expanduser("~/.bash_profile"),
            os.path.expanduser("~/.profile"),
            os.path.expanduser("~/.zshrc"),
            "/root/.bashrc",
            "/root/.bash_profile",
        ]:
            if os.path.isfile(shell_file):
                try:
                    for line in open(shell_file):
                        stripped = line.strip()
                        if stripped and not stripped.startswith("#"):
                            sev = _risk_score(stripped)
                            if sev in ("Critical", "High"):
                                findings.append({
                                    "type": "Shell Config",
                                    "location": shell_file,
                                    "name": os.path.basename(shell_file),
                                    "command": stripped,
                                    "severity": sev,
                                })
                except Exception:
                    pass

        # /etc/ld.so.preload — library injection
        if os.path.isfile("/etc/ld.so.preload"):
            try:
                content = open("/etc/ld.so.preload").read().strip()
                if content:
                    findings.append({
                        "type": "LD_PRELOAD Hijack",
                        "location": "/etc/ld.so.preload",
                        "name": "ld.so.preload",
                        "command": content,
                        "severity": "Critical",
                    })
            except Exception:
                pass

        return findings

    def _read_service_exec(self, path: str) -> str:
        try:
            for line in open(path):
                if line.strip().startswith("ExecStart="):
                    return line.strip()[len("ExecStart="):]
        except Exception:
            pass
        return "N/A"

File: agent-install/omni-agent-rs/src/capabilities/persistence_detection.rs
//! Persistence Detection capability — scans common persistence mechanisms.
//! Ports agent/capabilities/persistence_detection.py. Windows: registry Run
//! keys, Startup folders, scheduled tasks. Linux: systemd units, crontab,
//! /etc/cron.*, profile.d, shell rc files, /etc/ld.so.preload. Each finding is
//! risk-scored by suspicious command keywords (LOLBins, temp paths, encoders).

use super::Capability;
use serde_json::{json, Value};

const SUSPICIOUS_KEYWORDS: &[&str] = &[
    "powershell", "cmd.exe", "wscript", "cscript", "mshta", "rundll32",
    "regsvr32", "certutil", "bitsadmin", "curl", "wget", "nc ", "ncat",
    "python", "perl", "ruby", "base64", "encodedcommand", "-enc ",
    "/tmp/", "\\temp\\", "\\appdata\\local\\temp\\",
];

fn risk_score(command: &str) -> &'static str {
    let c = command.to_lowercase();
    let hits = SUSPICIOUS_KEYWORDS.iter().filter(|kw| c.contains(*kw)).count();
    if hits >= 2 {
        "Critical"
    } else if hits == 1 {
        "High"
    } else if c.contains("appdata") || c.contains("temp") {
        "Medium"
    } else {
        "Low"
    }
}

fn finding(kind: &str, location: &str, name: &str, command: &str, severity: &str) -> Value {
    json!({
        "type": kind,
        "location": location,
        "name": name,
        "command": command,
        "severity": severity,
    })
}

pub struct PersistenceDetectionCapability;

impl Capability for PersistenceDetectionCapability {
    fn id(&self) -> &'static str { "persistence_detection" }
    fn name(&self) -> &'static str { "Persistence Detection" }

    fn collect(&self, _sys: &sysinfo::System) -> Value {
        let findings = scan();
        let critical = findings.iter().filter(|f| f["severity"] == "Critical").count();
        let high = findings.iter().filter(|f| f["severity"] == "High").count();
        json!({
            "findings": findings,
            "count": findings.len(),
            "critical_count": critical,
            "high_count": high,
            "platform": if cfg!(windows) { "nt" } else { "posix" },
        })
    }
}

#[cfg(windows)]
fn scan() -> Vec<Value> {
    use winreg::enums::*;
    use winreg::RegKey;

    let mut findings: Vec<Value> = Vec::new();

    let run_keys: &[(&str, isize, &str)] = &[
        ("HKCU", HKEY_CURRENT_USER,  r"Software\Microsoft\Windows\CurrentVersion\Run"),
        ("HKCU", HKEY_CURRENT_USER,  r"Software\Microsoft\Windows\CurrentVersion\RunOnce"),
        ("HKCU", HKEY_CURRENT_USER,  r"Software\Microsoft\Windows\CurrentVersion\RunServices"),
        ("HKLM", HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows\CurrentVersion\Run"),
        ("HKLM", HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows\CurrentVersion\RunOnce"),
        ("HKLM", HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows\CurrentVersion\RunServices"),
        ("HKLM", HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows NT\CurrentVersion\Winlogon"),
        ("HKCU", HKEY_CURRENT_USER,  r"Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Run"),
        ("HKLM", HKEY_LOCAL_MACHINE, r"Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Run"),
    ];

    for (hive, hk, path) in run_keys {
        let root = RegKey::predef(*hk);
        if let Ok(key) = root.open_subkey(path) {
            for item in key.enum_values() {
                if let Ok((name, val)) = item {
                    let command = format!("{}", val);
                    let sev = risk_score(&command);
                    findings.push(finding(
                        "Registry Run Key",
                        &format!("{}\\{}", hive, path),
                        &name,
                        &command,
                        sev,
                    ));
                }
            }
        }
    }

    // Startup folders
    let mut startup_paths: Vec<String> = Vec::new();
    if let Ok(appdata) = std::env::var("APPDATA") {
        startup_paths.push(format!(r"{}\Microsoft\Windows\Start Menu\Programs\Startup", appdata));
    }
    startup_paths.push(r"C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Startup".to_string());
    for path in &startup_paths {
        if let Ok(entries) = std::fs::read_dir(path) {
            for entry in entries.flatten() {
                let item = entry.file_name().to_string_lossy().to_string();
                let full = entry.path().to_string_lossy().to_string();
                let sev = risk_score(&item);
                findings.push(finding("Startup Folder", path, &item, &full, sev));
            }
        }
    }

    // Scheduled tasks via schtasks
    findings.extend(scan_schtasks());
    findings
}

#[cfg(windows)]
fn scan_schtasks() -> Vec<Value> {
    let mut findings: Vec<Value> = Vec::new();
    let output = std::process::Command::new("schtasks")
        .args(["/query", "/fo", "CSV", "/v"])
        .output();
    let Ok(out) = output else { return findings };
    let text = String::from_utf8_lossy(&out.stdout);
    for line in text.lines() {
        // Skip the header row and non-data rows (mirrors the Python heuristic).
        if line.contains("TaskName") && line.contains("Task To Run") {
            continue;
        }
        let cols = parse_csv_line(line);
        if cols.len() >= 9 {
            let task_name = cols[0].trim();
            let task_cmd = cols[8].trim();
            if task_name.is_empty() || task_cmd.is_empty() || task_cmd.eq_ignore_ascii_case("Task To Run") {
                continue;
            }
            let sev = risk_score(task_cmd);
            if sev == "Critical" || sev == "High" || sev == "Medium" {
                findings.push(finding("Scheduled Task", "Windows Task Scheduler", task_name, task_cmd, sev));
            }
        }
    }
    findings
}

/// Minimal RFC-4180-ish splitter: handles double-quoted fields containing commas.
#[cfg(windows)]
fn parse_csv_line(line: &str) -> Vec<String> {
    let mut fields = Vec::new();
    let mut cur = String::new();
    let mut in_quotes = false;
    let mut chars = line.chars().peekable();
    while let Some(c) = chars.next() {
        match c {
            '"' if in_quotes && chars.peek() == Some(&'"') => { cur.push('"'); chars.next(); }
            '"' => in_quotes = !in_quotes,
            ',' if !in_quotes => { fields.push(std::mem::take(&mut cur)); }
            _ => cur.push(c),
        }
    }
    fields.push(cur);
    fields
}

#[cfg(not(windows))]
fn scan() -> Vec<Value> {
    let mut findings: Vec<Value> = Vec::new();

    // systemd services
    for dir in ["/etc/systemd/system", "/lib/systemd/system", "/usr/lib/systemd/system"] {
        if let Ok(entries) = std::fs::read_dir(dir) {
            for entry in entries.flatten() {
                let name = entry.file_name().to_string_lossy().to_string();
                if name.ends_with(".service") {
                    let full = entry.path().to_string_lossy().to_string();
                    let cmd = read_service_exec(&full);
                    let sev = risk_score(&cmd);
                    findings.push(finding("systemd Service", &full, &name, &cmd, sev));
                }
            }
        }
    }

    // user crontab
    if let Ok(out) = std::process::Command::new("crontab").arg("-l").output() {
        let text = String::from_utf8_lossy(&out.stdout);
        for line in text.lines() {
            let l = line.trim();
            if !l.is_empty() && !l.starts_with('#') {
                let sev = risk_score(l);
                findings.push(finding("Crontab Entry", "User Crontab", "Cron Job", l, sev));
            }
        }
    }

    // /etc/cron.d and /etc/cron.* directories
    for dir in ["/etc/cron.d", "/etc/cron.hourly", "/etc/cron.daily", "/etc/cron.weekly", "/etc/cron.monthly"] {
        if let Ok(entries) = std::fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_file() {
                    let item = entry.file_name().to_string_lossy().to_string();
                    if let Ok(content) = std::fs::read_to_string(&path) {
                        for line in content.lines() {
                            let l = line.trim();
                            if !l.is_empty() && !l.starts_with('#') {
                                let sev = risk_score(l);
                                findings.push(finding("Cron File", &path.to_string_lossy(), &item, l, sev));
                            }
                        }
                    }
                }
            }
        }
    }

    // /etc/profile.d scripts
    if let Ok(entries) = std::fs::read_dir("/etc/profile.d") {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                let item = entry.file_name().to_string_lossy().to_string();
                findings.push(finding("Profile Script", &path.to_string_lossy(), &item, "(loaded at login)", "Low"));
            }
        }
    }

    // User + root shell config files — only surface Critical/High lines
    let mut shell_files: Vec<String> = Vec::new();
    if let Ok(home) = std::env::var("HOME") {
        for f in [".bashrc", ".bash_profile", ".profile", ".zshrc"] {
            shell_files.push(format!("{}/{}", home, f));
        }
    }
    shell_files.push("/root/.bashrc".to_string());
    shell_files.push("/root/.bash_profile".to_string());
    for sf in &shell_files {
        if let Ok(content) = std::fs::read_to_string(sf) {
            for line in content.lines() {
                let l = line.trim();
                if !l.is_empty() && !l.starts_with('#') {
                    let sev = risk_score(l);
                    if sev == "Critical" || sev == "High" {
                        let base = std::path::Path::new(sf).file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
                        findings.push(finding("Shell Config", sf, &base, l, sev));
                    }
                }
            }
        }
    }

    // /etc/ld.so.preload — library injection
    if let Ok(content) = std::fs::read_to_string("/etc/ld.so.preload") {
        let c = content.trim();
        if !c.is_empty() {
            findings.push(finding("LD_PRELOAD Hijack", "/etc/ld.so.preload", "ld.so.preload", c, "Critical"));
        }
    }

    findings
}

#[cfg(not(windows))]
fn read_service_exec(path: &str) -> String {
    if let Ok(content) = std::fs::read_to_string(path) {
        for line in content.lines() {
            let l = line.trim();
            if let Some(rest) = l.strip_prefix("ExecStart=") {
                return rest.to_string();
            }
        }
    }
    "N/A".to_string()
}
